### ISSUE TITLE
Iterating on the V8 native binary format

### ISSUE THREAD

**Issue by titzer** [WebAssembly/design#497]
This public document (https://docs.google.com/a/google.com/document/d/1761v1AfhFM5kE8NArF_PyXcl-iVh0Dx3InOrmcyIoiI/pub) describes the V8 prototype binary format for WebAssembly.

This issue is to track discussions and enhancements possible to the format as motivated by design principles and pressures, data gathering, and performance.

**Comment by titzer** [comment]
I've made a publically accessible doc here: https://docs.google.com/document/d/1-G11CnMA0My20KI9D7dBR6ZCPOBCRD0oCH6SHCPFGx0/edit?usp=sharing

**Comment by AndrewScheidecker** [comment]
Thanks, very informative. Thoughts from my initial read:

1. There needs to be a backward compatible way to add value types. Maybe the signature table should be a type table that includes value types referenced by name.
2. Function body lengths should be more than 16-bits
3. There should be a separate export table from the function section. Making it simply a flag on the function doesn't support different export names or multiple exports of the same function.
4. Maybe there should be a separate import table from the function section. This is less clear, but I can see it being annoying to add e.g. module names or other import-specific data to the function section.
5. Locals shouldn't be a list of counts for a fixed set of types, since it isn't forward compatible with new value types
6. The global section isn't necessary. ASM.JS globals and thread-local memory can be implemented through data segments.
7. Data segments could have a flag for zero-initialization so dynamically linked modules can allocate zero-initialized static data.
8. What is the use case for the data segment init flag?
9. 16-bits isn't enough for function indices. UE4 ASM.JS has 120K functions, for example.

**Comment by JSStats** [comment]
Thank you for the write up.
* Memory sizes should be in pages or just bytes as they will not always be powers of two.
* Flatten the 'functions' section, effectively having one function per section - just make it a function definition section. With a size per section then functions could be individually skipped to dispatch processing to multiple threads. I would also like to add a text annotation section that could be included between sections this could then also handle text annotations between functions.
* Function bodies should remain inline, not elsewhere, for streaming etc.
* Global Variable section should not really be there, but seems ok if exploring thread local variables.
* Data segments should be page aligned in the binary for mapping into the address space? Interesting idea of on-demand loading of data segments.
* Do we really need integer constants of different bit widths or could a single LBE int constant suffice and flatten the i*const opcodes?
* Support inline strings rather than the current offsets as having the strings in a blob at the end and referenced by offset seems a show stopper for streaming.

**Comment by JSStats** [comment]
Sorry, taking back the suggestion to align the data segments as it might well be possible for the client to pad the file to suit as it streams in, or to store the data segments separately, etc.

**Comment by JSStats** [comment]
https://codereview.chromium.org/1565693002
WASM: Reserve an ignored section for source code meta information.

**Comment by kripken** [comment]
I'm implementing this in binaryen, and I might be missing something, but it looks like some things are missing, for example `f32.demote/f64`? (Should I file a separate issue? If so, where?)

**Comment by titzer** [comment]
They are named slightly differently. They are F64ConvertF32 and
F32ConvertF64, with the convention being the result type going first.

**Comment by kripken** [comment]
Hmm, I don't see that in the link in the 2nd comment? Is that document out of date, or not up to date perhaps?

**Comment by titzer** [comment]
You're right, it's not in the doc. I'll update that list.

**Comment by JSStats** [comment]
The design document, table 'Function Entry (repeated)' appears to have the order of the signature_index and the decl_flags backyards. Implementations appear to have the flags first.

**Comment by kripken** [comment]
Not sure if this is an issue or not. The spec says we write out functions one by one, each containing its body. Of course, the body of a function can contain a call to a function appearing *later*. In that case we need to parse that function's header before we can parse this function, because we don't know how to parse the `call` node - we don't know the function type, so we don't know how many arguments to read. So we need to wait for the other function.

In the worst case, many functions could call a single function at the very end. And then compiling them all would wait until the final one arrived. This seems like it defeats streaming parsing?

**Comment by JSStats** [comment]
@kripken Yes, decoding requires two passes. There is a note in the design document that suggests placing the bodies separately, and this might be enough for streaming. Would the function declarations alone be large and introduce much latency? Seems related to the suggestion of having a section map. If all the big blobs were ool then the high level structure and these declarations could all be parsed first, and then planning done on how to process the rest.

**Comment by kripken** [comment]
I see, thanks.

I suppose another option is to emit the function type in `call` and `call_import` nodes (`call_indirect` already does). At the cost of increasing those node sizes in the binary, of course.

**Comment by kg** [comment]
Emitting a signature value in every call would definitely enable streaming parsing, but it would bloat the file considerably. It's possible that gzip/brotli could recover most of the size, but it probably makes more sense to just split the function descriptions (the 'table') out from the bodies, so that the bodies can be at the end after all the descriptions. If you want quick streaming decode, you put the bodies at the end. Maybe it's better to put them at the end in general. I think it might be.

Once stream splitting is into the picture, there's not as much of an explicit ordering of anything, anyway - to do a streaming decode you will need to split the streams into chunks and interleave them. So streaming decode is a tricky problem we need to consider carefully.

**Comment by rossberg-chromium** [comment]
Just a reminder that annotating calls would _not_ remove the need for a second pass, since you'd still need to verify that the annotated signatures match the actual function signatures (which may come later).

AFAICS, the proper solution for streaming is to forward-declare all functions.

**Comment by lukewagner** [comment]
I've been thinking about this too recently.  In asm.js of course we don't have the forward declarations and so it's definitely something I've wished for.  However, it doesn't take two pasess: you just maintain a map from function-index to ("used with signature S" or "defined with signature S").  That's what we do in Odin (where signature is implied by the callsite).  This adds complexity, though, and is also troublesome if you want to do parallel verification (b/c of the mutability of the map) so I'd rather we just had the declaration table.

One funny thing is that an upfront table of function declarations inhibits streaming *generation* of wasm when you don't know the complete list of functions (or their signatures) up front.  However, I'm not sure if this is a problem we'd have in practice.  The main use cases for streaming wasm generation I'm thinking about are polyfills and user-space level 1 decompression and in both cases you can know this info up front or, in the case of polyfills, you can preemptively add all the helpers/shims you might possibly need to the declaration table and then make the bodies `unreachable` if they end up not being necessary.

**Comment by JSStats** [comment]
Another compromise option would be to include the signature index at each call site, so only the signatures need to be forward declared. The v8 encoding already requires the signatures before the code section. The signature index is already included with each indirect call.

Having the section layout early would help for other reasons. For example in augmenting the source code with information from another section it would help to at least be able to locate the position of the section without scanning through all the functions.

This issue could get very long if it's 'the' place to 'iterate' the v8 encoding to a consensus wasm encoding!

**Comment by JSStats** [comment]
@titzer Is there consensus to move the v8 encoding function bodies ool to unblock the streaming use case?

Proposal: Add a function_bodies_offset to the functions section. The function bodies would be read in turn from here. The body_size would remain in the functions section so that the layout of all functions would be known after reading the functions section alone. The intention is that a decoder would read all the sections up front and then decode the bodies, and the expectation is that all the sections are relatively small with bulk data ool.

**Comment by binji** [comment]
IIRC, in an older version of the v8 binary format, the function data was OOL. I can't remember why he moved it in-line.

**Comment by JSStats** [comment]
@titzer Another possible issue for streaming is the import and export function names being ool.  Are these needed to proceed with streaming compilation? Might some runtimes want to inline some imported functions or optimize the call sequence and thus need their names before compilation? Do people consider these too bulky to simply bundle into the sections read before beginning compilation, rather than ool and potentially near the end of the stream?

**Comment by lukewagner** [comment]
I was also wanting to propose moving exports/imports to the top of the module (probably in separate sections, right after the signature section), just waiting for there to be a more concrete proposal (in Markdown in this repo or ML in spec repo) to iterate on.

The reasoning for hoisting exports/imports is that these are what is necessary to build the module dependency graph that the browser module loader will load.  Putting these at the top allows the browser to fire off all the fetches earlier and in parallel.

**Comment by JSStats** [comment]
@lukewagner Good feedback, thank you. If as a small step the function import/export names are just moved inline rather than having them ool then would that unblock the issue? Effectively solving it for now with an experimental encoding change rather than spec change.

**Comment by JSStats** [comment]
@titzer Would there be consensus to change the min and max memory sizes from powers-of-two to bytes or a multiple of the page size?

**Comment by titzer** [comment]
On Fri, Jan 15, 2016 at 1:11 AM, JSStats <notifications@github.com> wrote:

> @titzer <https://github.com/titzer> Would there be consensus to change
> the min and max memory sizes from powers-of-two to bytes or a multiple of
> the page size?
>
Yep, multiple of page_size would be good.

**Comment by titzer** [comment]
On Thu, Jan 14, 2016 at 10:39 PM, JSStats <notifications@github.com> wrote:

> @titzer <https://github.com/titzer> Is there consensus to move the v8
> encoding function bodies ool to unblock the streaming use case?
>
> Proposal: Add a function_bodies_offset to the functions section. The
> function bodies would be read in turn from here. The body_size would remain
> in the functions section so that the layout of all functions would be known
> after reading the functions section alone. The intention is that a decoder
> would read all the sections up front and then decode the bodies, and the
> expectation is that all the sections are relatively small with bulk data
> ool.
>

Yeah, OOL function bodies sgtm. In fact, that was what I implemented in the
previous iteration of this format, around October. I changed them to be
inline because it was a little more convenient for test encoders. A simple
change would be to steal another bit from the decl_flags to indicate an OOL
function body is present.

**Comment by JSStats** [comment]
@titzer Thank you. Changing the memory sizes to pages will practically need wider fields, and u16 would be adequate for 32 bit but should we just use u32 sized fields.

Could we discuss a strategy for migrating to an incompatible encoding. For example, could a new section be added for the new memory sizes, with the old noted as deprecated, but both being accepted for a period? This could allow the tests and support code to migrate across after the new support has landed.

Similarly for the function section changes, could a new section be added so that a new  function_bodies_offset field could be added and all function bodies be ool in this new functions section, and the names inlined?

**Comment by kg** [comment]
> Could we discuss a strategy for migrating to an incompatible encoding. For example, could a new section be added for the new memory sizes, with the old noted as deprecated, but both being accepted for a period? This could allow the tests and support code to migrate across after the new support has landed.

I like the idea but I'm not sure whether it will work out in our favor. Deprecation is tricky to manage in an open/community project like this, and so are deadlines like the period you suggest. In some cases it sounds like we want to keep the old behavior alongside the 'right' behavior, like @titzer's suggestion above:

> A simple change would be to steal another bit from the decl_flags to indicate an OOL function body is present.

If we aren't thoughtful about this we can end up with a list of deprecated/optional behaviors where implementers have to make sense of which behaviors are preferred, which ones are actually used, and how they interact. It creates room for incompatible implementations down the road :-(

**Comment by JSStats** [comment]
There seems to be a show stopper in the interpretation of the function table index space. The v8 implementation excludes the imports from this index space. The binaryen implementation includes the imports claiming it is the correct interpretations of the document spec, see https://github.com/WebAssembly/binaryen/pull/115 Can people please work together to solve this.

**Comment by JSStats** [comment]
Similar show stopper for the order of the flags and signature index in the functions section. The design spec and the implementation do not agree. Binaryen is sticking with the design document spec. see https://github.com/WebAssembly/binaryen/pull/113 My suggestion if for people to follow the v8 implement except for clear bugs because being consistent with the implementation is what is required to 'get something working'.

**Comment by JSStats** [comment]
https://codereview.chromium.org/1597163002/
Issue 1597163002: wasm: change the module memory size to be multiples of the page size, 64k.

**Comment by titzer** [comment]
I think we should make the BinaryEncoding.md reflect the contents of the V8
document I sent out in December and update it to match the implementation.
I'll then "deprecate" the old doc and we can all work on keeping the
BinaryEncoding.md up to date.

**Comment by kg** [comment]
> I think we should make the BinaryEncoding.md reflect the contents of the V8 document I sent out in December and update it to match the implementation.

I've been working on that in my fork and will be submitting a PR when I'm done.

**Comment by flagxor** [comment]
We will close this now, deferring to individual issues.
